index.bs

```diff
@@ -1,164 +1,194 @@
 
-The above extension will allow reading values of locals or globals [[wasm-debugging-capabilities/#reconstruct-a-bindings-value|ᴿreconstruct-a-bindings-value]] (see [[#debug_info_stucture]] above). Compilers need to learn to track all mappings of the original source variables to operands stack locals, or globals (instead of native platform registers), calculate their lifetime and serialize that into `.debug_info` section.
+## DWARF Expressions and Location Descriptions ## {#DWARF-expressions-and-location-descriptions}
 
-This extension allows separating logically incompatible native platform concepts of registers/memory and WebAssembly specific elements. When DWARF information needed to be transformed for use in the native debugger, e.g. as JIT'ed code, the `DW_OP_WASM_location` extension can be easily replaced with native registers or memory location expression. There is no real advantage to overcomplicate logic of producers and consumers of the DWARF information with artificial mapping scheme that express WebAssembly specific location as CPU registers.
+<div class="note">
+Note: To enable the recovery of the values of variables, parameters, statics, etc... of a debuggee program at runtime, DWARF has *location descriptions* (see 2.6 of [[DWARF]]). There are four kinds of base, non-composite location description:
 
-### Proposed DWARF expression extension for locals ### {#dwarf-locals}
+1. Empty location descriptions (see 2.6.1.1.1 of [[DWARF]]) are used for optimized-away variables, or data that is otherwise unavailable.
 
-The `DW_OP_WASM_location` with code `0xED` can be added to the DWARF expression language. This operation will have two operands.
+2. Memory location descriptions (see 2.6.1.1.2 of [[DWARF]]) are used when a value is located at some address in memory.
 
-The first operand defines the type of the WebAssembly location. It is encoded as ULEB128 constant.
+3. Register location descriptions (see 2.6.1.1.3 of [[DWARF]]) are used when a value is located in a register.
 
-The second operand is a location item index and it is encoded as ULEB128 constant.
+4. Implicit location descriptions (see 2.6.1.1.4 of [[DWARF]]) are used when a value does not have any runtime representation, but has a known value anyways.
+
+Each of these location descriptions are applicable to values in WebAssembly, and may be used as they normally are, except for the third: register location descriptions. WebAssembly does not have registers per se. Instead, it has three distinct kinds of virtual registers (globals, locals, and the operand stack) and may use up to 2<sup>32</sup> - 1 instances of each virtual register.
+</div>
+
+When a program object's value is stored in a WebAssembly local, global, or on the operand stack it must be encoded as an offset from a well-known register number with the <code>DW_OP_breg<sub>n</sub></code> DWARF expression operation (see 2.5.1.2 of [[DWARF]]).
```

Comment:
Note: I think I'll revert `DW_OP_breg` back to custom `DW_OP_WASM_` for a reason of DW_OP_breg pointing to the memory address, which somewhat incompatible between wasm meaning and DWARF spec.

Reply:
Also, `DW_OP_breg*` exist because it is so common to have values at some offset from a register, and I expect it is equally common to have values exist at some offset from a local, or some offset from a global, and it might be worth having something like `DW_OP_wasm_blocal*` and `DW_OP_wasm_bglobal*` too. But this is also the kind of thing we can defer for now.